Bug 2005694: Removing proxy object takes up to 10 minutes for the changes to propagate to the MCO

jkyros opened the pull request:
This adjusts our behavior around how machine-config-operator reacts to config changes while the master pool has been paused. 

Old behavior on a config change with a paused master was: 
- `syncRequiredMachineConfigPools` spins for 10 minutes and times out with an error (which holds up other syncs such as RenderConfig) 
- Operator degrades after sync timeout error

New behavior is: 
- `syncRequiredMachineConfigPools` returns error immediately if pool is paused 
- Operator degrades after paused error

So we save the 10 minute timeout/wait that was useless anyway because we *knew* it would never succeed 

Fixes: https://bugzilla.redhat.com/show_bug.cgi?id=2005694

wking commented on pkg/operator/status.go:
```diff
@@ -308,6 +308,18 @@ func (optr *Operator) syncUpgradeableStatus() error {
+		// paused required pool will prevent us from upgrading
+		if pool.Spec.Paused {
+			if _, hasRequiredPoolLabel := pool.Labels[requiredForUpgradeMachineConfigPoolLabelKey]; hasRequiredPoolLabel {
+				coStatus.Status = configv1.ConditionFalse
```
Note that `Upgradeable=False` [only blocks minor updates][1] (4.y -> 4.(y+1)).  But :shrug:, if this is something we support, we can just document that folks should accept both "there will be no gates preventing you from triggering patch updates (4.y.z -> 4.y.z')" and "those patch updates will wedge waiting for the paused but required pool to complete".

[1]: https://github.com/openshift/api/blob/6c4f186339a7a4f505e133e9b86acd0d5aa84686/config/v1/types_cluster_operator.go#L184-L185

sinnykumari commented:
Would like to get thoughts from Jerry as well.
/cc @yuqi-zhang

yuqi-zhang commented:
Spoke with John about this, I think what we decided to do was to (mostly) preserve the existing behaviour by degrading immediately and not directly fiddling with upgradeable, since as Trevor points out, the upgradeable only blocks y stream upgrades.

As a longer term consideration, maybe we can rework how the syncs happen. For now, we will keep it as is (but degrade immediately for a paused master pool)

yuqi-zhang commented:
/test e2e-agnostic-upgrade

To see if I can trigger similar errors on non-drain related PRs

yuqi-zhang commented:
/test e2e-agnostic-upgrade

should be fixed now

jkyros commented:
I removed the "status" portion of this since we didn't want to do it now and updated descriptions accordingly -- I'll put some comments in [MCO-155](https://issues.redhat.com/browse/MCO-155) to capture what was going on here.

yuqi-zhang reviewed (approved):
LGTM!

In the future I think maybe we can rework these syncs to be a bit smarter, but for this bugfix, @jkyros and I discussed and we feel that preserving existing behaviour is a bit safer

sinnykumari commented:
/lgtm
